### DISCUSSION TITLE
How can I extract raw data at each stage?

### DISCUSSION THREAD

**Discussion by huang-kaixin** [The-OpenROAD-Project/OpenROAD-flow-scripts#1584] [Q&A]
Including Cell Density (density distribution of cells, equivalent to cell counts in each tile), the positions of macros and standard cells, and congestion. Like the raw data for the image below which generate in reports. Are there any corresponding scripts? Thank you for your response.
![final_placement](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/111551571/a1718fa4-518d-4dea-bc9b-e7d3d93d9fff)

**Comment by rovinski** [comment]
For cell density and congestion, check out the heat maps in the GUI. If these have the information you're looking for, you can dump it into a CSV file using [`gui::dump_heatmap`](https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/gui/README.md#heat-maps).

Macro positions can be obtained through [`write_macro_placement`](https://github.com/The-OpenROAD-Project/OpenROAD/blob/d9ecd4a7b3d486e0f33c8ec6a4b6677d360da8e7/src/mpl2/README.md#write-macro-placement), although this won't give you the raw data, it will give the data in the form of placement commands, so some simple post-processing would have to be done to get only the coordinates. Alternatively, you could extract the info yourself using Tcl and odb functions.

Dumping a list of all standard cell coordinates could be huge. There isn't a built-in function for it, but you could write a Tcl script to do it using odb functions.

**Comment by maliberty** [comment]
You can always use write_def to dump netlist information though you would have to parse that format.  Your other option would be to use the odb APIs to walk the netlist and write a custom dumper.
